File: java/Easy/_121_BestTimetoBuyandSellStock.java
package Easy;

import org.testng.annotations.Test;

public class _121_BestTimetoBuyandSellStock {
    public int maxProfit(int[] prices) {
        int res = 0;
        int minPrice = Integer.MAX_VALUE;

        for(int i = 0; i<prices.length; i++){
            if(prices[i] < minPrice) minPrice = prices[i];
            else res = Math.max(prices[i]-minPrice,res);
        }

        return res;
    }

    @Test
    public void test(){
        System.out.println(maxProfit(new int[]{7,1,5,3,6,4}));
    }
}

File: Python3/_121_BestTimeToBuyAndSellStock.py


from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left,right=0,1
        maxProfit = 0
        while right<len(prices):
            
            maxProfit = max(maxProfit, prices[right]-prices[left])
            if prices[left] < prices[right]:
                prices[right] = prices[left]
            left+=1
            right+=1
        return maxProfit


test = Solution()
test.maxProfit([7,1,5,3,6,4])